### ISSUE TITLE
Switch to newer Linux base container image by default

### ISSUE THREAD

**Issue by slonopotamus** [adamrehn/ue4-docker#296]
We're currently [defaulting to Ubuntu 18.04](https://github.com/adamrehn/ue4-docker/blob/v0.0.102/ue4docker/infrastructure/BuildConfiguration.py#L21-L22) whose main support cycle ends in April, 2023.

GitHub Actions, for example, is going to fully drop 18.04 support on 2023-04-01: https://github.com/actions/runner-images/issues/6002

I think we should change default to something newer.

**Comment by TBBle** [comment]
The main risks that come to mind:
* UE supported build platforms, particularly for older UE releases that are still supported.
* Building things against a newer libc that end up unexpectedly not runnable on older-libc releases that are still expected to be supported.

With Ubuntu 18.04 falling out of support, it absolutely makes sense to default to something upstream-supported, so the above should be taken as input on what to move to, not as against the move. I assume nothing will intrinsically block manually using an older Ubuntu release if that is necessary (legacy toolchain, old UE4 release, i386 support, etc.), although of course future evolution might have that effect (and can be addressed then if anyone turns out to care).

**Comment by slonopotamus** [comment]
> I assume nothing will intrinsically block manually using an older Ubuntu release if that is necessary

Sure, I'm only suggesting to change default value. If user wants to do something custom, they are absolutely free to do that.

**Comment by adamrehn** [comment]
> UE supported build platforms, particularly for older UE releases that are still supported.

There's a good chance that migrating to a newer version of Ubuntu will break Unreal Engine 4.20, which still had [version checking logic in Setup.sh](https://github.com/EpicGames/UnrealEngine/blob/4.20.3-release/Engine/Build/BatchFiles/Linux/Setup.sh) to install system packages for Mono. UE 4.21 and newer should theoretically be fine, since they bundle Mono (and later .NET Core), and [only ever try to install the `build-essential` system package](https://github.com/EpicGames/UnrealEngine/blob/4.21.2-release/Engine/Build/BatchFiles/Linux/Setup.sh), a behaviour that remains unchanged [as of the latest 5.1.1 release](https://github.com/EpicGames/UnrealEngine/blob/5.1.1-release/Engine/Build/BatchFiles/Linux/Setup.sh).

With this in mind, I would recommend that we drop official support for Unreal Engine 4.20 when we bump the default Ubuntu version, and update the README + docs to list 4.21 as the minimum supported version. Users who need to work with UE 4.20 can override the default value, or pin their version of ue4-docker to v0.0.103 to avoid any breaking changes that we may introduce in future releases when we no longer care about maintaining compatibility with 4.20.

> Building things against a newer libc that end up unexpectedly not runnable on older-libc releases that are still expected to be supported.

Fortunately, this is a non-issue for the Unreal Engine itself, thanks to the fact that it bundles a CentOS 7 sysroot alongside the clang compiler toolchain and links against the copy of glibc 2.17 from that. The only situation where people could potentially run into issues is if they're linking against system libraries from their projects or engine forks, and inadvertently introduce a transient dependency on a newer version of glibc as a result. In that scenario, I would argue that the onus is on the user to understand the implications of linking against system libraries, and that they should override the default with an appropriate value when building images to suit their requirements.

**Comment by adamrehn** [comment]
In addition to Ubuntu 18.04 reaching EOL, another reason to update has now presented itself: evidently Unreal Engine 5.2 crashes with Vulkan errors under Ubuntu 18.04, and [updating to 20.04 or newer fixes it](https://discord.com/channels/844118704941170688/918753427935346688/1131395829505216582). This will break scenarios where users run the Editor interactively from a container, whether via bind-mounting the host system's X11 socket or via Editor Pixel Streaming.

As per [this discussion in Discord](https://discord.com/channels/844118704941170688/918753427935346688/1131753554386882702), I'm gonna have a look at updating the default to a newer Ubuntu base image and verifying that it works with Unreal Engine 4.21 and newer. I figure we'll aim for Ubuntu 22.04, since that gives us glibc 2.35 and eliminates the need to [apply the DSO sorting patch](https://github.com/adamrehn/ue4-docker/blob/34b8e1a3226de572416dc9645597bb49bd0804a2/src/ue4docker/dockerfiles/ue4-build-prerequisites/linux/Dockerfile#L74-L83).

**Comment by slonopotamus** [comment]
@adamrehn any progress on adding Ubuntu 22.04 support?

**Comment by slonopotamus** [comment]
https://hub.docker.com/layers/nvidia/opengl/1.0-glvnd-devel-ubuntu22.04/images/sha256-c030a21588ddb89c245c8a2a24d6e68772cfa5939d312bab19f0833b90e9620a?context=explore

Oh. It turns out, Nvidia has published base image for 22.04. And I have evidence that someone already managed to build UE-5.2.0 with `-basetag ubuntu22.04`.

**Comment by adamrehn** [comment]
**Belated test results:** I ran tests for all supported versions of Unreal Engine (4.21 and newer), and everything seemed to work fine with the exception of Unreal Engine 5.0, which failed with the error `No usable version of libssl was found` due to its bundled .NET runtime looking for libssl 1.1 when Ubuntu 22.04 only ships with libssl 3.0 by default. This was a known compatibility issue for that release, and the [workarounds were awkward at best](https://github.com/MicrosoftDocs/live-share/issues/4646#issuecomment-1195426465).

Perhaps we should add something to the documentation warning users that they'll need to specify a different base image if they're building a 5.0.x release? In theory we could detect it and emit an error with a more helpful message, but the only way to reliably do so in the face of custom forks would be to parse `Build.version` after the engine source has already been cloned, which means it'd be erroring out at basically the exact same point in the build process anyway.

*(Side note: the fallout from the [changes Epic made to their CDN](https://forums.unrealengine.com/t/upcoming-disruption-of-service-impacting-unreal-engine-users-on-github/1155880) a while back certainly makes testing older Unreal Engine releases more annoying. I ended up creating branches for each release in a fork, and then committing the updated `Commit.gitdeps.xml` file for each branch.)*
